File: src/storage.ts
import { createClient } from 'redis';

import config from './config';

export const client = createClient({
  url: config.redisUrl,
});

const githubContributionsKey = (owner: string, repo: string) =>
  `github-contributions:${owner}:${repo}`;

export const storeToken = async (
  userId: string,
  accessToken: string,
  refreshToken: string
) => {
  await client.hSet(`user-tokens:${userId}`, { accessToken, refreshToken });
};

export const storeGitHubContributors = async (
  owner: string,
  repo: string,
  contributorIds: string[]
) => {
  const key = githubContributionsKey(owner, repo);
  await client
    .multi()
    .del(key)
    .sAdd(key, contributorIds)
    .expire(key, config.github.cacheSec, 'NX')
    .exec();
};

export const contributorsStored = async (owner: string, repo: string) => {
  const key = githubContributionsKey(owner, repo);
  return (await client.exists(key)) >= 1;
};

export const areContributors = async (
  owner: string,
  repo: string,
  contributorIds: string[]
) => {
  const key = githubContributionsKey(owner, repo);
  return (await client.smIsMember(key, contributorIds)).includes(true);
};

export const storeUserPlurality = async (userId: string) => {
  // Just store some value. We only care about the presence of this key
  await client
    .multi()
    .set(`user:${userId}:pk`, '0')
    .expire(`user:${userId}:pk`, 7 * 24 * 60 * 60)
    .exec();
};

export const isUserPlural = async (userId: string) => {
  return (await client.exists(`user:${userId}:pk`)) > 0;
};

export const connect = () => {
  client.connect();
};

File: src/storage.rs
use color_eyre::eyre::Result;
use log::*;
use poise::serenity_prelude::UserId;
use redis::{AsyncCommands as _, Client};

pub const USER_KEY: &str = "users-v1";

#[derive(Clone, Debug)]
pub struct Storage {
    client: Client,
}

impl Storage {
    pub fn new(redis_url: &str) -> Result<Self> {
        let client = Client::open(redis_url)?;

        Ok(Self { client })
    }

    pub async fn store_user_plurality(&self, sender: UserId) -> Result<()> {
        let mut con = self.client.get_async_connection().await?;

        info!("Marking {sender} as a PluralKit user");
        let key = format!("{USER_KEY}:{sender}:pk");

        // Just store some value. We only care about the presence of this key
        con.set(&key, 0).await?;
        con.expire(key, 7 * 24 * 60 * 60).await?;

        Ok(())
    }

    pub async fn is_user_plural(&self, user_id: UserId) -> Result<bool> {
        let key = format!("{USER_KEY}:{user_id}:pk");
        let mut con = self.client.get_async_connection().await?;

        let exists: bool = con.exists(key).await?;
        Ok(exists)
    }
}
